File: backend/src/main/java/by/fpmibsu/findafriend/controller/controllers/AdvertController.java
package by.fpmibsu.findafriend.controller.controllers;

import by.fpmibsu.findafriend.application.HandleResult;
import by.fpmibsu.findafriend.application.controller.*;
import by.fpmibsu.findafriend.application.mediatr.Mediatr;
import by.fpmibsu.findafriend.controller.AuthUtils;
import by.fpmibsu.findafriend.controller.Logging;
import by.fpmibsu.findafriend.controller.Validation;
import by.fpmibsu.findafriend.controller.commands.adverts.CreateAdvertCommand;
import by.fpmibsu.findafriend.controller.commands.adverts.DeleteAdvertCommand;
import by.fpmibsu.findafriend.controller.commands.adverts.UpdateAdvertCommand;
import by.fpmibsu.findafriend.controller.queries.adverts.GetAdvertByIdQuery;
import by.fpmibsu.findafriend.controller.queries.adverts.GetAdvertsQuery;
import by.fpmibsu.findafriend.dataaccesslayer.advert.AdvertDao;
import by.fpmibsu.findafriend.entity.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@ControllerRoute(route = "/adverts")
public class AdvertController extends Controller {
    private final Logger logger = LogManager.getLogger(AdvertController.class);
    private final Mediatr mediatr;

    public AdvertController(Mediatr mediatr) {
        this.mediatr = mediatr;
    }

    @Endpoint(path = "/getAll", method = HttpMethod.GET)
    public HandleResult getAll() throws Exception {
        return ok(mediatr.send(new GetAdvertsQuery()));
    }

    @Endpoint(path = "/getById", method = HttpMethod.GET)
    public HandleResult getById(@FromQuery(parameterName = "id") int id) throws Exception {
        return ok(mediatr.send(new GetAdvertByIdQuery(id)));
    }

    @RequireAuthentication
    @Endpoint(path = "/create", method = HttpMethod.POST)
    public HandleResult createAdvert(@FromBody CreateAdvertCommand request) throws Exception {
        if (Validation.isAnyNullOrEmpty(request.description, request.title)
                || (request.advertType < 0 || request.advertType > 1)) {
            return badRequest();
        }
        return ok(mediatr.send(request));
    }

    @RequireAuthentication
    @Endpoint(path = "/update", method = HttpMethod.PUT)
    public HandleResult updateAdvert(@FromBody UpdateAdvertCommand request, @WebToken(parameterName = "id") int userId,
                                     @WebToken(parameterName = "role") String role) throws Exception {
        if (Validation.isAnyNullOrEmpty(request.description, request.title)) {
            return badRequest();
        }
        if (!AuthUtils.allowRoles(role, User.Role.ADMINISTRATOR, User.Role.MODERATOR)) {
            var advertDao = serviceProvider.getRequiredService(AdvertDao.class);
            var advert = advertDao.getEntityById(request.advertId);
            if (advert == null) {
                return badRequest();
            }
            if (advert.getOwner().getId() != userId) {
                Logging.warnNonAuthorizedAccess(this.request, logger);
                return notAuthorized();
            }
        }
        return ok(mediatr.send(request));
    }

    @RequireAuthentication
    @Endpoint(path = "/delete", method = HttpMethod.DELETE)
    public HandleResult deleteAdvertById(@FromQuery(parameterName = "id") int id,
                                         @WebToken(parameterName = "id") int userId,
                                         @WebToken(parameterName = "role") String role) throws Exception {
        if (!AuthUtils.allowRoles(role, User.Role.ADMINISTRATOR, User.Role.MODERATOR)) {
            var advertDao = serviceProvider.getRequiredService(AdvertDao.class);
            var advert = advertDao.getEntityById(id);
            if (advert == null) {
                return badRequest();
            }
            if (advert.getOwner().getId() != userId) {
                Logging.warnNonAuthorizedAccess(this.request, logger);
                return notAuthorized();
            }
        }
        logger.trace("response is completed");
        return ok(mediatr.send(new DeleteAdvertCommand(id)));
    }
}

File: BackendNET/Backend/Controllers/AdvertController.cs
using Backend.Commands.Adverts.CreateAdvert;
using Backend.Commands.Adverts.DeleteAdvert;
using Backend.Commands.Adverts.UpdateAdvert;
using Backend.Dto;
using Backend.Queries.Adverts.GetAll;
using Backend.Queries.Adverts.GetById;
using Backend.Services;
using Backend.Utils;
using MediatR;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using static Backend.Entities.User;

namespace Backend.Controllers;

public class AdvertController : ApiController
{
    private readonly IMediator _mediator;

    public AdvertController(IMediator mediator)
    {
        _mediator = mediator;
    }

    [HttpGet("/adverts/getAll")]
    public async Task<ActionResult<IEnumerable<AdvertDto>>> GetAllAsync()
    {
        return Ok(await _mediator.Send(new GetAllAdvertsQuery()));
    }

    [HttpGet("/adverts/getById")]
    public async Task<ActionResult<AdvertDto?>> GetByIdAsync([FromQuery] int id)
    {
        return Ok(await _mediator.Send(new GetAdvertByIdQuery { AdvertId = id }));
    }

    [Authorize]
    [HttpPost("/adverts/create")]
    public async Task<IActionResult> CreateAsync([FromBody] CreateAdvertCommand command)
    {
        command.UserId = AuthData.UserId;
        return Ok(await _mediator.Send(command));
    }

    [Authorize]
    [HttpPut("/adverts/update")]
    public async Task<IActionResult> UpdateAsync([FromBody] UpdateAdvertCommand command)
    {
        if (string.IsNullOrEmpty(command.Title) || string.IsNullOrEmpty(command.Description))
        {
            return BadRequest();
        }

        if (AuthData.UserRole is not UserRole.Administrator and not UserRole.Moderator)
        {
            var userId = AuthData.UserId;
            if (!await Service<IAdvertService>().IsOwnerOfAdvertAsync(userId, command.AdvertId, CancellationToken))
            {
                return Unauthorized();
            }
        }
        return Ok(await _mediator.Send(command));
    }

    [Authorize]
    [HttpDelete("/adverts/delete")]
    public async Task<IActionResult> DeleteAsync([FromQuery] int id)
    {
        if (AuthData.UserRole is not UserRole.Administrator and not UserRole.Moderator)
        {
            var userId = AuthData.UserId;
            if (!await Service<IAdvertService>().IsOwnerOfAdvertAsync(userId, id, CancellationToken))
            {
                return Unauthorized();
            }
        }
        return Ok(await _mediator.Send(new DeleteAdvertCommand { AdvertId = id }));
    }
}